lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -0,0 +1,377 @@
+  // Get the execution state for the run.
+  auto it = executionStates_.find(runId);
+
+  if (it == executionStates_.end()) {
+    assert("Execution state should exist!");
+  }
+
+  std::shared_ptr<ExecutionState> executionState = it->second;
+
+  // Get the map of Placeholder -> Tensor mappings within ctx.
+  const Context::PlaceholderMap &phMap = ctx->pairs();
+
+  // Get the symbol table for the node.
+  const RuntimeBundle &bundle = node->runtimeBundle;
+  const SymbolTable &symbolTable = bundle.getSymbolTable();
+
+  for (const auto &symbolPair : symbolTable) {
+    const auto &symbolName = symbolPair.first;
+    const auto &symbolInfo = symbolPair.second;
+
+    if (symbolInfo.input) {
+      // If the symbol is an input, look for a Placeholder in ctx with
+      // the same name and copy the corresponding Tensor into the input Context
+      // being prepared for the node.
+      for (const auto &phTensorPair : phMap) {
+        auto *placeholder = phTensorPair.first;
+        const auto *tensor = phTensorPair.second;
+
+        if (symbolName == placeholder->getName()) {
+          executionState->insertIntoNodeCtx(node, placeholder, tensor->clone());
+        }
+      }
+    }
+  }
+}
+
+void ThreadPoolExecutor::executeDAGNode(RunIdentifierTy runId,
+                                        const DAGNode *node) {
+  // Get the execution state for the run.
+  auto executionStateIt = executionStates_.find(runId);
+
+  if (executionStateIt == executionStates_.end()) {
+    assert("Execution state should exist!");
+  }
+
+  std::shared_ptr<ExecutionState> executionState = executionStateIt->second;
+
+  // If execution has already failed due to another node, don't bother running
+  // this one.
+  if (executionState->getResultCode() == ResultCode::Failed) {
+    // Mark the node as no longer execution.
+    executionState->decrementInflightNodes();
+    return;
+  }
+
+  // Get the DeviceManager that can run the node.
+  auto deviceManagerIt = deviceManagers_.find(node->deviceID);
+
+  if (deviceManagerIt == deviceManagers_.end()) {
+    // Mark the node as no longer executing.
+    executionState->decrementInflightNodes();
+    executionState->setResultCode(ResultCode::Failed);
+    return;
+  }
+
+  std::shared_ptr<DeviceManager> deviceManager = deviceManagerIt->second;
+
+  // Get the Context containing all of the inputs for the node.
+  std::unique_ptr<Context> nodeCtx =
+      executionState->getUniqueNodeContextPtr(node);
+  // Run the node using the DeviceManager.
+  deviceManager->runFunction(
+      node->name, std::move(nodeCtx),
+      [this, runId, node](RunIdentifierTy id, ResultCode resultCode,
+                          std::unique_ptr<Context> resultCtx) {
+        // Immediately move the handling of the result onto threadPool_ to
+        // avoid doing work on the DeviceManager thread.
+        std::shared_ptr<Context> sharedResultCtx = std::move(resultCtx);
```

Comment:
Why do you convert to shared_ptr here? If you want to use shared_ptr instead of unique_ptr we should talk about making that the interface to the DeviceManager.

An advantage of using unique_ptr is that you can read/write the Context safely without locks because you know you don't have any other pointers to it. With shared_ptr I'm pretty suspicious of accesses to Context::pairs() for example.

Reply:
I did this because I wasn't able to capture the `unique_ptr` in the lambda passed to `threadPool_` below :(. I read something online about how anything you want to capture in a lambda has to be moveable AND copyable, but this might have changed for newer C++ versions?

Reply:
oh you need to do a move alias in the lambda capture, something like this:
```
[ctx = std::move(resultCtx), ...]
```